README
NAME
    Catalyst::Action::FromPSGI - Use a PSGI app as a Catalyst action

VERSION
    version 0.001005

SYNOPSIS
    First, you have a psgi app you wrote and want to use:

     package MyApp::WS::App;

     use Web::Simple;

     has name => (
        is => 'ro',
        required => 1,
     );

     sub dispatch_request {
        sub (/hi) {
           [ 200,
              [ 'Content-type' => 'text/plain' ],
              [ 'Hello ' . $_[0]->name ]
           ]
        },
     }

     1;

    Now you want to reuse this app in a Catalyst action:

     package MyApp::Controller::HelloName;

     use base 'Catalyst::Controller';

     sub say_hi :Path('/say_hi_to') ActionClass('FromPSGI') {
       my ($self, $c, $name, @args) = @_;

       MyApp::WS::App->new(name => $name)->to_psgi_app
     }

     1;

    The above would yield 'Hello fREW' for the request to
    "/say_hi_to/fREW/hi".

    Of course the above example is contrived, but keep in mind this will
    work for any of the myriad PSGI apps out there.

DESCRIPTION
    "Catalyst::Action::FromPSGI" gives you a handy way to mount PSGI apps
    under Catalyst actions.

    Note that because Catalyst is in control of the dispatch cycle any
    limitations you place on it will be placed on the PSGI app as well. So
    for example:

     sub foo : Path('/foo') Args(1) ActionClass('FromPSGI') { ... }

    will never run the PSGI app if the url is "/foo/bar/baz" because the
    Catalyst dispatcher won't even match for more than one argument. For
    this reason I recommend leaving "Args" unspecified for "FromPSGI"
    actions.

    I actually made this because I'm interested in using Web::Machine
    instead of Catalyst::Action::REST and possibly even replacing my
    chaining code with Web::Simple based dispatching.

THANKS
    Matt S. Trout - for pioneering the actual guts of this code. Stevan
    Little - for porting Web::Machine, my motivation for making this.

AUTHOR
    Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Arthur Axel "fREW" Schmidt.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Catalyst-Action-FromPSGI

0.001005  2013-09-18 15:56:55 America/Chicago
  - Fixed yet another psgi.input issue (thanks Wes Malone for finding this)

0.001004  2013-07-12 23:41:13 CST6CDT
  - Correctly set psgi.input (thanks Fitz Elliott for the test/bug report and
    miyagawa for helping find the fix)

0.001003  2013-06-13 19:15:03 CST6CDT
  - abandon RT
  - match trailing slashes when nesting (Fitz Elliott)

0.001002  2013-05-13 14:26:54 CST6CDT
  - fix terminology (PSGI is spec, Plack is impl)
  - fix warning from Plack::App::URLMap

0.001001  2012-11-17 16:43:57 CST6CDT
  - Put MetaYAML back in dist

0.001000  2012-06-24 17:48:55 America/Chicago
  - Initial Release



Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Use a PSGI app as a Catalyst action",
  "AUTHOR" => "Arthur Axel \"fREW\" Schmidt <frioux+cpan\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Catalyst-Action-FromPSGI",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Catalyst::Action::FromPSGI",
  "PREREQ_PM" => {
    "Catalyst::Runtime" => "5.90013",
    "Plack" => "0.992"
  },
  "TEST_REQUIRES" => {
    "HTTP::Request::Common" => 0,
    "Test::More" => "0.88",
    "Web::Simple" => "0.016"
  },
  "VERSION" => "0.001005",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Catalyst/Action/FromPSGI.pm
package Catalyst::Action::FromPSGI;
{
  $Catalyst::Action::FromPSGI::VERSION = '0.001005';
}

# ABSTRACT: Use a PSGI app as a Catalyst action

use strict;
use warnings;
use base 'Catalyst::Action';
use HTTP::Message::PSGI qw(res_from_psgi);
use Plack::App::URLMap;
use Plack::Request;

sub nest_app {
   my ($self, $c, $app) = @_;

   my $nest = Plack::App::URLMap->new;

   my $path = '/' . $c->request->path;
   my $rest = join '/', @{$c->request->arguments};
   $path =~ s/\Q$rest\E\/?$//;
   $nest->map( $path => $app );

   return $nest->to_app
}

sub snort_psgi {
   my ($self, $c, $r) = @_;

   $c->res->status($r->code);
   $c->res->body($r->content);
   $c->res->headers($r->headers);
}

sub execute {
   my ($self, $controller, $c, @rest) = @_;

   my $app = $self->code->($controller, $c, @rest);
   my $nest = $self->nest_app($c, $app);

   my $body = Plack::Request->new($c->req->env)->body;
   $c->req->env->{'psgi.input'} = ref $body
      ? do { $body->seek(0, 0); $body }
      : do { open my $fh, '<', \$body; $fh }
   ;
   $c->req->env->{'psgix.input.buffered'} = 1;

   my $res = res_from_psgi($nest->($c->req->env));
   $self->snort_psgi($c, $res);

   return;
}

1;

__END__

=pod

=head1 NAME

Catalyst::Action::FromPSGI - Use a PSGI app as a Catalyst action

=head1 VERSION

version 0.001005

=head1 SYNOPSIS

First, you have a psgi app you wrote and want to use:

 package MyApp::WS::App;

 use Web::Simple;

 has name => (
    is => 'ro',
    required => 1,
 );

 sub dispatch_request {
    sub (/hi) {
       [ 200,
          [ 'Content-type' => 'text/plain' ],
          [ 'Hello ' . $_[0]->name ]
       ]
    },
 }

 1;

Now you want to reuse this app in a Catalyst action:

 package MyApp::Controller::HelloName;

 use base 'Catalyst::Controller';

 sub say_hi :Path('/say_hi_to') ActionClass('FromPSGI') {
   my ($self, $c, $name, @args) = @_;

   MyApp::WS::App->new(name => $name)->to_psgi_app
 }

 1;

The above would yield C<'Hello fREW'> for the request to
C</say_hi_to/fREW/hi>.

Of course the above example is contrived, but keep in mind this will work for
any of the myriad PSGI apps out there.

=head1 DESCRIPTION

C<Catalyst::Action::FromPSGI> gives you a handy way to mount PSGI apps
under Catalyst actions.

Note that because Catalyst is in control of the dispatch cycle any limitations
you place on it will be placed on the PSGI app as well.  So for example:

 sub foo : Path('/foo') Args(1) ActionClass('FromPSGI') { ... }

will never run the PSGI app if the url is C</foo/bar/baz> because the
Catalyst dispatcher won't even match for more than one argument.  For this
reason I recommend leaving C<Args> unspecified for C<FromPSGI> actions.

I actually made this because I'm interested in using L<Web::Machine> instead
of L<Catalyst::Action::REST> and possibly even replacing my chaining code
with L<Web::Simple> based dispatching.

=head1 THANKS

Matt S. Trout - for pioneering the actual guts of this code.  Stevan Little -
for porting L<Web::Machine>, my motivation for making this.

=head1 AUTHOR

Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Arthur Axel "fREW" Schmidt.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
